### PR TITLE
feat: add metadata column type badge

### PR DESCRIPTION
**Pull request by thiagodallacqua-hpe**
## Ticket

[ET-791](https://hpe-aiatscale.atlassian.net/browse/ET-791)

## Description

Right now, if we have multiple metadata columns with the same name, there's no quick way of knowing of which type each one is while selecting it. This PR intends to add a descriptive badge next to duplicate entries for the metadata columns.

## Test Plan

- make sure that you have at least one metadata column that is duplicated and of a different type
- go to a experiment list
  - make sure that the grid visualization is enabled
- open the column picker
- verify that the duplicates has a `<column_name> <column_type_tag>` format

Screenshot 2024-10-14 at 16 58 13

Screenshot 2024-10-14 at 16 59 27

Screenshot 2024-10-21 at 11 12 49

Screenshot 2024-10-21 at 11 12 42

Screenshot 2024-10-21 at 11 13 43

Screenshot 2024-10-18 at 17 44 26

## Checklist

- [ ] Changes have been manually QA'd
- [ ] New features have been approved by the corresponding PM
- [ ] User-facing API changes have the "User-facing API Change" label
- [ ] Release notes have been added as a separate file under `docs/release-notes/`
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses have been included for new code which was copied and/or modified from any external code

[ET-790]: https://hpe-aiatscale.atlassian.net/browse/ET-790?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[DET-123]: https://hpe-aiatscale.atlassian.net/browse/DET-123?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

[ET-791]: https://hpe-aiatscale.atlassian.net/browse/ET-791?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by keita-determined — CHANGES_REQUESTED** [review]
we wanna show types for all columns.
and keys need to be adjusted.

**Comment by keita-determined** [comment]
can we prioritize this PR over the PR about filter?

File: webui/react/src/components/ColumnPickerMenu.tsx [hunk 1]
```diff
@@ -95,18 +96,44 @@ const ColumnPickerTab: React.FC<ColumnTabProps> = ({
   }, [searchString, totalColumns, tab]);
 
   const allFilteredColumnsChecked = useMemo(() => {
-    return filteredColumns.every((col) => columnState.includes(col.column));
+    return filteredColumns.every((col) => {
+      const colType = col.type.replace('COLUMN_TYPE_', '');
+
+      if (col.column.includes('metadata'))
+        return columnState.includes(col.column.concat(`_${colType}`));
```
Comment (ashtonG) [inline webui/react/src/components/ColumnPickerMenu.tsx:103]: please read the epic ticket. we need to be careful about how we encode the column name because the backend does not place any limitations about the column name, so if the user includes a type name in the column name, this could become problematic.

use the method discussed in the ticket to encode the type/column name tuple.

File: webui/react/src/components/ColumnPickerMenu.tsx [hunk 2]
```diff
@@ -165,26 +197,63 @@ const ColumnPickerTab: React.FC<ColumnTabProps> = ({
   const rows = useCallback(
     ({ index, style }: { index: number; style: React.CSSProperties }) => {
       const col = filteredColumns[index];
+      const colType = col.type.replace('COLUMN_TYPE_', '');
+      const getColDisplayName = (col: ProjectColumn) => {
+        const metCol = metadataColumns.get(col.column);
+
+        if (metCol !== undefined && metCol.length > 1) {
```
Comment (ashtonG) [inline webui/react/src/components/ColumnPickerMenu.tsx:204]: if we're always including the badge, do we still need the column count state and this branch?

**Review by ashtonG — CHANGES_REQUESTED** [review]

File: webui/react/src/components/ColumnPickerMenu.tsx [hunk 3]
```diff
@@ -21,6 +22,8 @@ const BANNED_COLUMNS: Set<string> = new Set([]);
 const removeBannedColumns = (columns: ProjectColumn[]) =>
   columns.filter((col) => !BANNED_COLUMNS.has(col.column));
 
+export const METADATA_SEPARATOR = '\u241F' as const;
```
Comment (keita-determined) [inline webui/react/src/components/ColumnPickerMenu.tsx:25]: can we move this [here](https://github.com/determined-ai/determined/pull/10066/files#diff-b3f8c84db4390025a5db99526f13b9320f76cd069d83a295ab3c7ec59810d3b5)

File: webui/react/src/components/ColumnPickerMenu.tsx [hunk 4]
```diff
@@ -63,6 +66,12 @@ interface ColumnTabProps {
   onHeatmapSelectionRemove?: (id: string) => void;
 }
 
+export const formatColumnKey = (col: ProjectColumn): string => {
```
Comment (keita-determined) [inline webui/react/src/components/ColumnPickerMenu.tsx:69]: i think we wanna store all columns with types based on the discussion 2 weeks ago.

**Review by keita-determined — CHANGES_REQUESTED** [review]
Could you make sure this PR displays the columns' data correctly on #10066? 
so far, it doesnt show data correctly.

**Review by ashtonG — CHANGES_REQUESTED** [review]
found a bug -- with multiple types of the same metadata column selected, changing the column width of one column changes the width of all columns with the same name. please make sure the column ids that we're sending to the datagrid component are correct.

File: webui/react/src/components/ColumnPickerMenu.tsx (hunk 4 above)
Comment (ashtonG) [reply webui/react/src/components/ColumnPickerMenu.tsx:69]: just confirming -- as the only affected columns are run metadata columns, we should be okay to limit the type encoding to run metadata columns.

File: webui/react/src/components/ColumnPickerMenu.tsx [hunk 5]
```diff
@@ -165,24 +188,44 @@ const ColumnPickerTab: React.FC<ColumnTabProps> = ({
...
     ({ index, style }: { index: number; style: React.CSSProperties }) => {
       const col = filteredColumns[index];
+      const colType = col.type.replace('COLUMN_TYPE_', '');
+      const getColDisplayName = (col: ProjectColumn) => {
+        return (
+          <>
+            {col.displayName || col.column} <Badge text={colType} />
+          </>
+        );
+      };
```
Comment (ashtonG) [inline webui/react/src/components/ColumnPickerMenu.tsx:198]: please make sure the "known unspecified" columns (archived, parent archived, part of search) are badged as boolean to prevent confusion

File: webui/react/src/components/ColumnPickerMenu.tsx [hunk 6]
```diff
@@ -124,31 +138,35 @@ const ColumnPickerTab: React.FC<ColumnTabProps> = ({
   const handleColumnChange = useCallback(
     (event: CheckboxChangeEvent) => {
       const { id, checked } = event.target;
+
       if (id === undefined) return;
+
+      const targetCol = id;
```
Comment (ashtonG) [inline webui/react/src/components/ColumnPickerMenu.tsx:144]: is this reassignment necessary?

File: webui/react/src/components/ColumnPickerMenu.tsx [hunk 7]
```diff
@@ -95,18 +99,28 @@ const ColumnPickerTab: React.FC<ColumnTabProps> = ({
...
     const newColumns = allFilteredColumnsChecked
       ? columnState.filter((col) => !filteredColumnMap[col])
-      : [...new Set([...columnState, ...filteredColumns.map((col) => col.column)])];
+      : [
+          ...new Set([
+            ...columnState,
+            ...filteredColumns.map((col) => {
+              if (col.location === V1LocationType.RUNMETADATA) return formatColumnKey(col);
+              return col.column;
+            }),
```
Comment (ashtonG) [inline webui/react/src/components/ColumnPickerMenu.tsx:121]: `formatColumnKey` already does this check
```suggestion
            ...filteredColumns.map((col) => formatColumnKey(col)),
```

File: webui/react/src/components/ColumnPickerMenu.tsx [hunk 8]
```diff
@@ -95,18 +99,28 @@ const ColumnPickerTab: React.FC<ColumnTabProps> = ({
...
   const handleShowHideAll = useCallback(() => {
-    const filteredColumnMap: Record<string, boolean> = filteredColumns.reduce(
-      (acc, col) => ({ ...acc, [col.column]: columnState.includes(col.column) }),
-      {},
-    );
+    const filteredColumnMap: Record<string, boolean> = filteredColumns.reduce((acc, col) => {
+      if (col.location === V1LocationType.RUNMETADATA)
+        return { ...acc, [formatColumnKey(col)]: columnState.includes(formatColumnKey(col)) };
+
+      return { ...acc, [col.column]: columnState.includes(col.column) };
```
Comment (ashtonG) [inline webui/react/src/components/ColumnPickerMenu.tsx:110]: see prev re: `formatColumnKey`
```suggestion
      return { ...acc, [formatColumnKey(col)]: columnState.includes(col.column) };
```

File: webui/react/src/components/Searches/Searches.tsx [hunk 9]
```diff
@@ -676,6 +676,7 @@ const Searches: React.FC<Props> = ({ project }) => {
               currentColumn.displayName || currentColumn.column,
               settings.columnWidths[currentColumn.column],
               dataPath,
+              undefined,
```
Comment (ashtonG) [inline webui/react/src/components/Searches/Searches.tsx:679]: shouldn't you be able to remove this argument without issue? `heatmapProps` is marked as optional and behind datapath.

File: webui/react/src/utils/flatRun.ts [hunk 10]
```diff
@@ -116,3 +117,17 @@ export const getIdsFilter = (
     filterGroup,
   };
 };
+
+export const removeColumnTypePrefix = (columnName: V1ColumnType): string => {
+  return columnName.replace('COLUMN_TYPE_', '');
+};
+
+/// wanna know why this separator is used? see https://hpe-aiatscale.atlassian.net/browse/ET-785
+export const COLUMN_SEPARATOR = '␟';
```
Comment (ashtonG) [inline webui/react/src/utils/flatRun.ts:126]: this doesn't seem to be in use.

File: webui/react/src/components/ColumnPickerMenu.tsx (hunk 4 above)
Comment (keita-determined) [reply webui/react/src/components/ColumnPickerMenu.tsx]: As far as i know, hyperparameter would have the same names with different types too. i think its safer to encode all columns for the future and consistency.

File: webui/react/package-lock.json
Comment (keita-determined) [inline webui/react/package-lock.json:1]: is this change related to this PR?
